Trying to remove false positive on logical or.

Neirda24 opened the pull request:
This PR:

- [x] Improve LogicalOr mutatability
- [x] Covered by tests

<!--
Remove if not relevant
-->

Fixes https://github.com/infection/infection/issues/1796

It improves the LogicalOr mutatability by detecting if we are trying to mutate to something that doesn't make sense.
For example :

```php
$myVar === 'hello' || $myVar === 'world';
```

cannot be mutated to 

```php
$myVar === 'hello' && $myVar === 'world';
```

I think there a lot more usecase that I don't cover such as

```php
$myVar < 5 || $myVar > 10;
```

can't be mutated but

```php
$myVar < 10 || $myVar > 5;
```

could. So not only should we check the variable name but also to what it compares. Although it would require to handle a lot of cases and Node class doesn't help much at the moment. Help appreciated to improve this further.

---
Edit : also handled some use cases regarding < & >

Neirda24 commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,72 @@ public function mutate(Node $node): iterable
+            $varName = '';
+
+            if ($node->left->left instanceof Node\Expr\Variable) {
+                $varName = $node->left->left->name;
```
for some reasons, phpstan and phpstorm cannot infer the type Variable to `$node->left->left` although it is checked against the instanceof right before...

Neirda24 commented on tests/phpunit/Mutator/Boolean/LogicalOrTest.php:
```diff
@@ -66,11 +66,62 @@ public function mutationsProvider(): iterable
+            <<<'PHP'
+<?php
+
+$myVar < 10 || $myVar > 5;
```
This should trigger a mutation because of the value. Not in the current scope of this PR ATM.

Neirda24 commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,72 @@ public function mutate(Node $node): iterable
+
+        if (
+            (
+                in_array(get_class($node->left), $greaterOp, true) === true
```
Could have applied this logic only once but I figured if we need to compare the values too then the logic might differ

maks-rafalko commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,76 @@ public function mutate(Node $node): iterable
+        ];
+
+        if (
+            in_array(get_class($node->left), $equalOp, true) === true
```
`get_class($node->left)` - does it make sense to move to variable? As it's being called many times. Same for `$node->right`

Neirda24 replied:
Yes we could. When upgrading PHP we could then use `$node->left::class` and maybe we will have to rollback the variable.

maks-rafalko replied:
ok, let's do it

Neirda24 replied:
done

maks-rafalko commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,76 @@ public function mutate(Node $node): iterable
+
+            if ($node->left->left instanceof Node\Expr\Variable) {
+                $varNameLeft = $node->left->left->name;
+            } elseif ($node->left->right instanceof Node\Expr\Variable) {
```
could you please explain what source code leads to this checks/AST? I think I don't understand AST structure :)

Neirda24 replied:
😅 
```php
$myVar === 'hello' || $myVar === 'world';
```

for example (see added tests). Because it must not be mutated to 

```php
$myVar === 'hello' && $myVar === 'world';
```

So I check that on both side of the `||` we compare the same variable name against an equality. but because the variable could be on either the right side of `===` or the left side, I have to check node->left->left, node->left->right + node->right->right, node->left->left  for the variable name.

maks-rafalko replied:
cool, now I fully understand it. Then IMO in tests we need to add a case where on the left side of `||` we have a variable on the right side of `===` (now there is no such case)

Neirda24 replied:
yes there is : https://github.com/infection/infection/pull/1801/files#diff-6cb62609cbceaf882b2803a9e26b277dc505d2cae4258e29d23acf9a9d067d37R106

maks-rafalko replied:
I mean on the left side of `||`. So additional test will be

```diff
+ 'hello' === $myVar || 'world' === $myVar;
```

Neirda24 replied:
fixed.

Neirda24 commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,76 @@ public function mutate(Node $node): iterable
+        ];
+
+        if (
+            in_array(get_class($node->left), $equalOp, true) === true
```
As you wish. I always add a strict comparison because sometimes we want it to be equal to false and the `!` is easy to miss / remove. I'll adapt to your coding style.

maks-rafalko commented on src/Mutator/Boolean/LogicalOr.php:
```diff
@@ -77,6 +79,79 @@ public function mutate(Node $node): iterable
+        if (
+            (
+                in_array($classNodeLeft, $greaterOp, true) === true
+                && in_array($classNodeRight, $smallerOp, true) === true
```
still present, please check all cases :)

maks-rafalko commented:
could you also please fix PHPStan issues? https://github.com/infection/infection/actions/runs/3985550828/jobs/6833052618#step:7:90

Neirda24 commented:
@maks-rafalko : sorry for the delay. I've figured that some of what I was trying to fix shouldn't be. I'll update the PR with added tests soon !

Neirda24 commented:
@maks-rafalko : Okay I've pushed new code and fixed coding style using cs. I'll gadly take some help finishing this as it requires I think a lot of tests combinations to reduce the number of mutants. Some of those mutants can't be tested against I think. Let me know.

maks-rafalko commented:
@Neirda24  could you please rebase on master?

Neirda24 commented:
@maks-rafalko : here are the result of psalm :

```bash
$ docker compose run php81 make psalm
./.tools/psalm --threads=max

Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///opt/infection/.tools/psalm/.box/src/RequirementCollection.php on line 15

Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///opt/infection/.tools/psalm/.box/src/RequirementCollection.php on line 19
Target PHP version: 8.0 (inferred from composer.json)
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 411 (14%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 411 (29%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 411 (43%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 240 / 411 (58%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 300 / 411 (72%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 360 / 411 (87%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░

ERROR: UndefinedPropertyFetch - src/Mutator/Boolean/LogicalOr.php:100:25 - Instance property PhpParser\Node\Expr::$left is not defined (see https://psalm.dev/039)
        $nodeLeftLeft = $nodeLeft->left;


ERROR: UndefinedPropertyFetch - src/Mutator/Boolean/LogicalOr.php:101:26 - Instance property PhpParser\Node\Expr::$right is not defined (see https://psalm.dev/039)
        $nodeLeftRight = $nodeLeft->right;


ERROR: UndefinedPropertyFetch - src/Mutator/Boolean/LogicalOr.php:103:26 - Instance property PhpParser\Node\Expr::$left is not defined (see https://psalm.dev/039)
        $nodeRightLeft = $nodeRight->left;


ERROR: UndefinedPropertyFetch - src/Mutator/Boolean/LogicalOr.php:104:27 - Instance property PhpParser\Node\Expr::$right is not defined (see https://psalm.dev/039)
        $nodeRightRight = $nodeRight->right;


ERROR: UndefinedMethod - src/Mutator/Boolean/LogicalOr.php:202:41 - Method PhpParser\Node\Expr::getOperatorSigil does not exist (see https://psalm.dev/022)
            return (match ("{$nodeLeft->getOperatorSigil()}::{$nodeRight->getOperatorSigil()}") {


ERROR: UndefinedMethod - src/Mutator/Boolean/LogicalOr.php:202:75 - Method PhpParser\Node\Expr::getOperatorSigil does not exist (see https://psalm.dev/022)
            return (match ("{$nodeLeft->getOperatorSigil()}::{$nodeRight->getOperatorSigil()}") {


------------------------------
6 errors found
------------------------------
416 other issues found.
You can display them with --show-info=true
------------------------------
Psalm can automatically fix 5 of these issues.
Run Psalm again with
--alter --issues=InvalidNullableReturnType,InvalidReturnType --dry-run
to see what it can fix.
------------------------------
```

But regarding `ERROR: UndefinedPropertyFetch - src/Mutator/Boolean/LogicalOr.php:100:25 - Instance property PhpParser\Node\Expr::$left is not defined (see https://psalm.dev/039)
        $nodeLeftLeft = $nodeLeft->left;` for example the issue is that right before here is the code :

```php
        if (
            !property_exists($nodeLeft, 'left')
            || !property_exists($nodeLeft, 'right')
            || !property_exists($nodeRight, 'left')
            || !property_exists($nodeRight, 'right')
        ) {
            return true;
        }

        $nodeLeftLeft = $nodeLeft->left;
        $nodeLeftRight = $nodeLeft->right;

        $nodeRightLeft = $nodeRight->left;
        $nodeRightRight = $nodeRight->right;
```

looks like a psalm issue don't you think ? There is no Interface that ensure there is a left / right property.

Neirda24 commented:
@maks-rafalko : Fixed psalm but it looks ugly.... Let me know what you think

maks-rafalko commented:
> @maks-rafalko : Fixed psalm but it looks ugly.... Let me know what you think

yeah, no the best look, but I can't advice anything better, so I'm personally ok with that.

That's already a huge job! Could you look into PHPStan error as well and make `Autoreview` build green?

> Some of those mutants can't be tested against I think. Let me know.

If they can't be tested, probably we can ignore them?

Neirda24 commented:
@maks-rafalko : there is an open issue about this (I thought I saw it merged with an article but couldn't find it) https://github.com/phpstan/phpstan/issues/4451 . Basically PHPStan complains that there is no `default` but in this case the default would be an execption same as the default behavior without the `default` key. There is still an option to set teh default to true (In doubt mutate) otherwise. What do you prefer ? It seems to be the only thing left to fix (I'll also look into mutations in CI to see if I can fix more)

maks-rafalko commented:
Probably exception will be better (if this is the same as default behavior when `default` is absent?

> I'll also look into mutations in CI to see if I can fix more

👍

Neirda24 commented:
Ok I think I fixed everything I could. I added more tests and fixed psalm. Although introducing the default in match statement fixes psalm, it creates a new mutation that we cannot test (unreachable statement). Mutation still failing are Either LogicalOr | LogicalAnd or InstanceOf and I can't seems to be able to test them directly. I think we are dealing with other false positive here. Could you check and let me know what you think please @maks-rafalko  ? Also I re-arramged a bit the tests.

maks-rafalko commented:
I've fixed CS and added ignore rules.

Thank you for your contribution @Neirda24

Neirda24 commented:
Thank you for the final touch !
